packages/headless/src/api/commerce/commerce-api-params.ts

```diff
@@ -26,25 +26,36 @@ export interface ContextParam {
 export interface ContextParams {
   view: ViewParams;
   user?: UserParams;
-  cart?: CartParams[];
-  labels?: Record<string, string>;
+  cart?: CartItemParam[];
 }
 
 export interface ViewParams {
   url: string;
 }
 
-export interface UserParams {
-  userAgent?: string;
-  userIp?: string;
+interface UserIdRequired {
+  userId: string;
   email?: string;
+}
+
+interface EmailRequired {
   userId?: string;
+  email: string;
 }
 
-export interface CartParams {
-  groupId?: string;
-  productId?: string;
-  sku?: string;
+interface UserIdAndEmail {
+  userId: string;
+  email: string;
+}
+
+export type UserParams = (UserIdRequired | EmailRequired | UserIdAndEmail) & {
```

Comment:
> See https://github.com/coveo/ui-kit/pull/3190#discussion_r1337184673
> We want to require at least a user id, or at least an email, or both.

I understand that part, but won't we need to have an if-condition within the class/method logic itself to ensure that at least one of the fields is present in the object in order to appease the type check? If it's possible for _either field to be undefined_ without us knowing which field _will be defined_, we'll have to rely on an if-condition to ensure at least one is defined to appease the type check anyways. And if we need to do that, it's functionally the same as using a Partial of the `UserIdAndEmail` interface, right?

Alternatively, if we know that one of the fields will always be populated within a certain context based on its type (e.g., email is populated by a method call that returns a string), then such an if-condition isn't necessary and it still passes the type-check.

This will be my last comment arguing this, I promise 😅 If I'm overlooking something that you're already accounting for, ignore this comment and carry on. I just wanted to elaborate on my thought-process and why I'm uncertain about this being necessary.

Reply:
The type definition is enough to ensure we do not need to do an if/else to ensure it's valid! Plus, the user is sent pretty much as-is to the API; we don't need to manipulate it much.

However, you are right that we'll need an `if` condition to do validation (in the slice): https://github.com/coveo/ui-kit/pull/3190/files#diff-21dff18bbe200c24fe0d214536833d65408d08b0f38b3e26b1d1546da94132e3R36

> And if we need to do that, it's functionally the same as using a Partial of the UserIdAndEmail interface, right?

Not sure I get this. By having the type defined as-is, for typescript users of headless, we can enforce the requirement. For javascript users, we have the condition as defined above.